### PR TITLE
Add option to log subprocess output to files in DDP launcher.

### LINKED ISSUE #7134: torch.distributed.launch outputs all processes to stdout (single node)
Using the new `torch.distributed.launch` function on a single node, we get multiple processes outputting to the same stdout, which leads to a messy screen!

Previous solution from `multiproc.py` via the NVIDIA guys used stdout for the main process, and then outputted to log files for the remainder of the processes. Unsure what would be the cleanest way to handle this, but willing to help!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

### PR DESCRIPTION
**Pull request by rohan-varma**
Closes https://github.com/pytorch/pytorch/issues/7134. This request is to add an option to log the subprocess output (each subprocess is training a network with DDP) to a file instead of the default stdout.

The reason for this is that if we have N processes all writing to stdout, it'll be hard to decipher the output, and it would be cleaner to log these to separate files.

To support this, we add an optional argument `--logdir` set the subprocess stdout to be the a file of the format "node_rank_{}_local_rank_{}" in the logging directory. With this enabled, none of the training processes output to the parent process stdout, and instead write to the aformentioned file. If a user accidently passes in something that's not a directory, we fallback to ignoring this argument.

Tested by taking a training script at https://gist.github.com/rohan-varma/2ff1d6051440d2c18e96fe57904b55d9 and running `python -m torch.distributed.launch --nproc_per_node=2 --nnodes=1 --node_rank=0 --master_addr="127.0.0.1" --master_port="29500" --logdir test_logdir train.py`. This results in a directory `test_logdir` with files "node_0_local_rank_0" and "node_0_local_rank_1" being created with the training process stdout.

### CODE REVIEW

**Comment by rohan-varma** [comment]
Will add reviewers after polishing

**Comment by glenn-jocher** [comment]
@rohan-varma is there any official pytorch solution for this? We are running into this problem in YOLOv5, and are attempting to manually patch together a messy python logger solution here https://github.com/ultralytics/yolov5/pull/719, but it would be ideal if there were a more general solution for the wider community.

**Comment by hendrytl** [comment]
@rohan-varma What is the status of this PR?  I would love to take advantage of this capability.

**Comment by rohan-varma** [comment]
Thanks for checking in @glenn-jocher @hendrytl! I will clean up this PR and aim to publish it for review this week.

**Comment by glenn-jocher** [comment]
@rohan-varma awesome, thank you! This will help a lot of YOLOv5 users using DDP multi-gpu :)

File: torch/distributed/launch.py [hunk 1]
```diff
@@ -249,14 +268,25 @@ def main():
 
         cmd.extend(args.training_script_args)
 
-        process = subprocess.Popen(cmd, env=current_env)
+        subprocess_stdout = None
+        if args.logdir:
+            directory_path = os.path.join(os.getcwd(), args.logdir)
+            file_handle = open(os.path.join(directory_path, "process_{}".format(local_rank)), "w")
```
Comment (hendrytl) [inline torch/distributed/launch.py:274]: Would it make sense to also include the node_rank so that log file names from multiple machines do not collide?  Perhaps "process_{}_{}".format(node_rank, local_rank)?

**Comment by rohan-varma** [comment]
@glenn-jocher @hendrytl I updated the PR, feel free to take a look and let me know if it works for your use case. I included an example in the PR description.

File: torch/distributed/launch.py [hunk 2]
```diff
@@ -255,6 +289,10 @@ def main():
             raise subprocess.CalledProcessError(returncode=process.returncode,
                                                 cmd=cmd)
 
+    # close open file descriptors
+    for file_handle in subprocess_file_handles:
```
Comment (hendrytl) [inline torch/distributed/launch.py:293]: Add to a finally block to ensure file handles are closed if an error is raised.

**Review by hendrytl — APPROVED** [review]

**Review by glenn-jocher — APPROVED** [review]

**Comment by rohan-varma** [comment]
@pritamdamania87 @mrshenli Would be great if you can take a look at this as it seems like it would be a useful addition for users of `torch.distributed.launch`.

**Comment by kiukchung** [comment]
fwiw we were planning to add this exact feature to the torchelastic agent (and naturally we’d have to add it to torchelastic.distributed.launch).

File: torch/distributed/launch.py [hunk 3]
```diff
@@ -220,6 +231,18 @@ def main():
               "your application as needed. \n"
               "*****************************************".format(current_env["OMP_NUM_THREADS"]))
 
+    if args.logdir:
+        # Possibly create the directory to write subprocess log output to.
+        if os.path.exists(args.logdir):
+            if not os.path.isdir(args.logdir):
+                print("passed in --logdir must be a relative path to a directory. Ignoring argument.")
+                args.logdir = None
```
Comment (pritamdamania87) [inline torch/distributed/launch.py:239]: We should probably raise an Error here instead of continuing without logging.

File: torch/distributed/launch.py [hunk 4]
```diff
@@ -246,14 +269,31 @@ def main():
...
+            node_rank = args.node_rank
+            stdout_file_name = node_local_rank_stdout_filename.format(node_rank, local_rank)
+            file_path = os.path.join(directory_path, stdout_file_name)
+            file_handle = open(file_path, "w")
+            subprocess_stdout = file_handle
+            subprocess_file_handles.append(file_handle)
+            print(f"Note: Stdout for node {node_rank} rank {local_rank} will be written to {file_path}")
+
+
+        process = subprocess.Popen(cmd, env=current_env, stdout=subprocess_stdout)
```
Comment (pritamdamania87) [inline torch/distributed/launch.py:284]: Shouldn't we also write stderr to a file?

Comment (rohan-varma) [reply torch/distributed/launch.py]: Updated the diff to do this.

File: torch/distributed/launch.py [hunk 5]
```diff
@@ -235,7 +236,7 @@ def main():
         # Possibly create the directory to write subprocess log output to.
         if os.path.exists(args.logdir):
             if not os.path.isdir(args.logdir):
-                print("passed in --logdir must be a relative path to a directory. Ignoring argument.")
+                raise ValueError("argument --logdir must be a path to a directory.")
                 args.logdir = None
```
Comment (pritamdamania87) [inline torch/distributed/launch.py:240]: This will be never executed and can be removed now?

**Review by pritamdamania87 — APPROVED** [review]
